File: Scala/data_structures/src/main/scala-2.12/heaps_and_sets/Heapify.scala
package heaps_and_sets

object Heapify {
  import scala.annotation.tailrec
  type IndexPair = (Int, Int)

  private def convertToIntArray(line: String): Array[Int] = line.split(" ").map(_.toInt)

  private def getIndexOfParentChildrenMinimum(array: Array[Int], parentIndex: Int, size: Int): Int = {
    val leftChildIx: Int = 2 * parentIndex + 1
    val rightChildIx: Int = leftChildIx + 1
    val minIndex: Int = if (leftChildIx < size &&  array(leftChildIx) < array(parentIndex)) leftChildIx else parentIndex
    if (rightChildIx < size && array(rightChildIx) < array(minIndex)) rightChildIx else minIndex
  }

  private[heaps_and_sets] def swapElems(array: Array[Int], ix: Int, jy: Int): Unit = {
    val elem: Int = array(ix)
    array(ix) = array(jy)
    array(jy) = elem
  }

  private def siftDown(array: Array[Int], parentIndex: Int, n: Int): List[IndexPair] = {
    @tailrec
    def loop(currentParentIx: Int, currentMinIx: Int, swaps: List[IndexPair]): List[IndexPair] = {
      if (currentMinIx == currentParentIx) swaps
      else {
        swapElems(array, currentMinIx, currentParentIx)
        val nextParentIx: Int = currentMinIx
        val nextMinIx: Int = getIndexOfParentChildrenMinimum(array, nextParentIx, n)
        loop(nextParentIx, nextMinIx, (currentParentIx, currentMinIx) :: swaps)
      }
    }
    val minIndex: Int = getIndexOfParentChildrenMinimum(array, parentIndex, n)
    loop(parentIndex, minIndex, Nil)
  }

  def heapify(array: Array[Int], n: Int): List[IndexPair] = {
    @tailrec
    def loop(swaps: List[IndexPair], parentIndex: Int): List[IndexPair] = {
      if (parentIndex < 0) swaps.reverse
      else {
        val currentSwaps: List[IndexPair] = siftDown(array, parentIndex, n)
        loop(currentSwaps ::: swaps, parentIndex - 1)
      }
    }
    loop(Nil, n / 2 - 1)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val n: Int = reader.next().toInt
    val array: Array[Int] = convertToIntArray(reader.next())
    val swaps: List[IndexPair] = heapify(array, n)
    println(swaps.length)
    swaps.foreach{ case (i, j) => println(s"$i $j")}
  }
}

File: OCaml/data_structures/heapify.ml
let convert_to_intlist line = List.map int_of_string Str.(line |> split (regexp " "))


let get_index_of_parent_children_minimum arr parent_index size =
    let left_child_ix = 2 * parent_index + 1 in
    let right_child_ix = left_child_ix + 1 in
    let min_index =
        if left_child_ix < size && arr.(left_child_ix) < arr.(parent_index) then left_child_ix
        else parent_index in
    if right_child_ix < size && arr.(right_child_ix) < arr.(min_index) then right_child_ix else min_index


let swap_elems arr ix jy =
    let elem = arr.(ix) in
    arr.(ix) <- arr.(jy);
    arr.(jy) <- elem


let sift_down arr parent_ix n =
    let rec loop current_parent_ix current_min_ix swaps =
        if current_min_ix = current_parent_ix then swaps
        else
            let () = swap_elems arr current_min_ix current_parent_ix in
            let next_parent_ix = current_min_ix in
            let next_min_ix = get_index_of_parent_children_minimum arr next_parent_ix n in
            loop next_parent_ix next_min_ix ((current_parent_ix, current_min_ix) :: swaps) in
    let min_index = get_index_of_parent_children_minimum arr parent_ix n
    in loop parent_ix min_index []


let heapify arr n =
    let rec loop swaps parent_index =
        if parent_index < 0 then List.rev swaps
        else
            let current_swaps = sift_down arr parent_index n in
            loop (current_swaps @ swaps) (parent_index - 1)
    in loop [] (n / 2  - 1)


let () =
    let n = read_int() in
    let arr = Array.of_list @@ convert_to_intlist (read_line()) in
    let swaps = heapify arr n in
    print_int (List.length swaps);
    print_newline();
    List.iter (function (i, j) -> print_endline (string_of_int i ^ " " ^ string_of_int j)) swaps
